libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+     * @param hostName Hostname to be used to form the QnAMaker host URL, which follows the following format https://{hostName}.azurewebsites.net/qnamaker
+     * @param noAnswer (Optional) Activity to be sent in the event no answer is found within the knowledgebase.
+     * @param threshold (Optional) The threshold above which to treat answers found from the knowledgebase as a match.
+     * @param activeLearningCardTitle (Optional) Title of the card displayed showing active learning options if active learning is enabled.
+     * @param cardNoMatchText (Optional) Text to be show on a button alongside active learning options, allowing a user to indicate none of the options are applicable.
+     * @param top (Optional) Maximum number of answers to return from the knowledgebase.
+     * @param cardNoMatchResponse (Optional) Activity to be sent if the user selects the no match option on an active learning card.
+     * @param strictFilters (Optional) QnAMakerMetadata collection used to filter / boost queries to the knowledgebase.
+     * @param dialogId (Optional) Id of the created dialog. Default is 'QnAMakerDialog'.
+     */
+    constructor(knowledgeBaseId, endpointKey, hostName, noAnswer = null, threshold = 0.3, activeLearningCardTitle = 'Did you mean:', cardNoMatchText = 'None of the above.', top = 3, cardNoMatchResponse: Activity = null, strictFilters: QnAMakerMetadata[] = null, dialogId = 'QnAMakerDialog') {
+        super(dialogId);
+
+        this.knowledgeBaseId = knowledgeBaseId;
+        this.hostName = hostName;
+        this.endpointKey = endpointKey;
+        this.threshold = threshold;
+        this.top = top;
+        this.activeLearningCardTitle = activeLearningCardTitle;
+        this.cardNoMatchText = cardNoMatchText;
+        this.strintFilters = strictFilters;
+        this.noAnswer = noAnswer;
+        this.cardNoMatchResponse = cardNoMatchResponse;
+
+        this.addStep(this.callGenerateAnswer.bind(this));
+        this.addStep(this.callTrain.bind(this));
+        this.addStep(this.checkForMultiTurnPrompt.bind(this));
+        this.addStep(this.displayQnAResult.bind(this));
+    }
+
+    public async beginDialog(dc: DialogContext, options?: object): Promise<DialogTurnResult> {
+        if (!dc) { throw new Error('Missing DialogContext'); }
+        
+        if (dc.context.activity.type != ActivityTypes.Message) 
+        { 
+            return Dialog.EndOfTurn; 
+        }
+
+        const dialogOptions: QnAMakerDialogOptions = {
+            qnaDialogResponseOptions: await this.getQnAResponseOptions(),
+            qnaMakerOptions: await this.getQnAMakerOptions()
+        };
+
+        if(options)
+        {
+            Object.assign(dialogOptions, options);
+        }
+
+        return await super.beginDialog(dc, dialogOptions);
+    }
+
+    /**
+     * Returns a new instance of QnAMakerOptions.
+    **/
+    private async getQnAMakerOptions(): Promise<QnAMakerOptions> {
```

Comment:
>async [](start = 12, length = 5)

why is this async?

Reply:
Agreed - it shouldn't be. Altered. This was a hangover from the C# version which should arguably be changed, but not sure if that breaks binary compat?